### PR TITLE
SpectrumInfo rollout: algorithms 43 to 47

### LINKED ISSUE #18539: SpectrumInfo rollout 43 to 47
This issue is part of the SpectrumInfo rollout, see #17743.
Algorithms covered in this issue are:

- Framework/Crystal/src/AnvredCorrection.cpp
- Framework/Crystal/src/NormaliseVanadium.cpp
- Framework/CurveFitting/src/Algorithms/ConvertToYSpace.cpp
- Framework/CurveFitting/src/Functions/FunctionQDepends.cpp
- Framework/CurveFitting/src/Functions/VesuvioResolution.cpp
- Framework/CurveFitting/src/Functions/ComptonPeakProfile.cpp

### PR DESCRIPTION
**Pull request by PranavBahuguna**
This PR is part of refactoring effort to use the new interface of the SpectrumInfo module.
See #17743 for details.

The following algorithms were refactored to use SpectrumInfo:

- Framework/Crystal/src/AnvredCorrection.cpp
- Framework/Crystal/src/NormaliseVanadium.cpp
- Framework/CurveFitting/src/Algorithms/ConvertToYSpace.cpp
- Framework/CurveFitting/src/Functions/FunctionQDepends.cpp
- Framework/CurveFitting/src/Functions/VesuvioResolution.cpp
- Framework/CurveFitting/src/Functions/ComptonPeakProfile.cpp
- Framework/CurveFitting/src/Algorithms/VesuvioCalculateGammaBackground.cpp

Performance tests were added for:
- Framework/Crystal/src/NormaliseVanadium.cpp
- Framework/CurveFitting/src/Algorithms/ConvertToYSpace.cpp

None of the algorithms have significantly improved performance.

Fixes #18539.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Crystal/src/AnvredCorrection.cpp [hunk 1]
```diff
@@ -168,30 +166,21 @@ void AnvredCorrection::exec() {
...
-    if (!det)
+    // If no detector is found, skip onto the next spectrum
+    if (!spectrumInfo.hasDetectors(i))
       continue;
 
     // This is the scattered beam direction
     Instrument_const_sptr inst = m_inputWS->getInstrument();
-    V3D dir = det->getPos() - samplePos;
+    V3D dir = spectrumInfo.position(i) - samplePos;
     double L2 = dir.norm();
```
Comment (SimonHeybrock) [inline Framework/Crystal/src/AnvredCorrection.cpp]: `spectrumInfo.l2(i)`

File: Framework/Crystal/src/AnvredCorrection.cpp [hunk 2]
```diff
@@ -168,30 +166,21 @@ void AnvredCorrection::exec() {
...
     // Two-theta = polar angle = scattering angle = between +Z vector and the
     // scattered beam
```
Comment (SimonHeybrock) [inline Framework/Crystal/src/AnvredCorrection.cpp]: `double scattering = spectrumInfo.twoTheta(i)`

File: Framework/Crystal/src/AnvredCorrection.cpp [hunk 3]
```diff
@@ -168,30 +166,21 @@ void AnvredCorrection::exec() {
...
-      // in an openmp block.
-    }
-    // If no detector found, skip onto the next spectrum
-    if (!det)
+    // If no detector is found, skip onto the next spectrum
+    if (!spectrumInfo.hasDetectors(i))
       continue;
 
     // This is the scattered beam direction
     Instrument_const_sptr inst = m_inputWS->getInstrument();
```
Comment (SimonHeybrock) [inline Framework/Crystal/src/AnvredCorrection.cpp]: Now unused?

File: Framework/Crystal/src/AnvredCorrection.cpp [hunk 4]
```diff
@@ -280,6 +269,8 @@ void AnvredCorrection::execEvent() {
   const V3D pos = inst->getSource()->getPos() - samplePos;
   double L1 = pos.norm();
```
Comment (SimonHeybrock) [inline Framework/Crystal/src/AnvredCorrection.cpp]: `spectrumInfo.l1()`

File: Framework/Crystal/src/AnvredCorrection.cpp [hunk 5]
```diff
@@ -168,30 +166,21 @@ void AnvredCorrection::exec() {
   const V3D pos = m_inputWS->getInstrument()->getSource()->getPos() - samplePos;
   double L1 = pos.norm();
```
Comment (SimonHeybrock) [inline Framework/Crystal/src/AnvredCorrection.cpp]: `spectrumInfo.l1()`

File: Framework/Crystal/src/AnvredCorrection.cpp [hunk 6]
```diff
@@ -289,23 +280,12 @@ void AnvredCorrection::execEvent() {
...
-    // If no detector found, skip onto the next spectrum
-    if (!det)
+    // If no detector is found, skip onto the next spectrum
+    if (!spectrumInfo.hasDetectors(i))
       continue;
 
     // This is the scattered beam direction
-    V3D dir = det->getPos() - samplePos;
+    V3D dir = spectrumInfo.position(i) - samplePos;
     double L2 = dir.norm();
```
Comment (SimonHeybrock) [inline Framework/Crystal/src/AnvredCorrection.cpp]: `spectrumInfo.l2(i)`

File: Framework/Crystal/src/AnvredCorrection.cpp [hunk 7]
```diff
@@ -289,23 +280,12 @@ void AnvredCorrection::execEvent() {
...
     // Two-theta = polar angle = scattering angle = between +Z vector and the
     // scattered beam
```
Comment (SimonHeybrock) [inline Framework/Crystal/src/AnvredCorrection.cpp]: `spectrumInfo.twoTheta`

File: Framework/Crystal/src/NormaliseVanadium.cpp [hunk 8]
```diff
@@ -63,6 +62,8 @@ void NormaliseVanadium::exec() {
   const V3D pos = m_inputWS->getInstrument()->getSource()->getPos() - samplePos;
   double L1 = pos.norm();
```
Comment (SimonHeybrock) [inline Framework/Crystal/src/NormaliseVanadium.cpp]: `spectrumInfo.l1()`

File: Framework/Crystal/src/NormaliseVanadium.cpp [hunk 9]
```diff
@@ -79,24 +80,13 @@ void NormaliseVanadium::exec() {
...
-    if (!det)
+    // If no detector is found, skip onto the next spectrum
+    if (!spectrumInfo.hasDetectors(i))
       continue;
 
     // This is the scattered beam direction
     Instrument_const_sptr inst = m_inputWS->getInstrument();
-    V3D dir = det->getPos() - samplePos;
+    V3D dir = spectrumInfo.position(i) - samplePos;
     double L2 = dir.norm();
```
Comment (SimonHeybrock) [inline Framework/Crystal/src/NormaliseVanadium.cpp]: `spectrumInfo.l2(i)`

File: Framework/Crystal/src/NormaliseVanadium.cpp [hunk 10]
```diff
@@ -79,24 +80,13 @@ void NormaliseVanadium::exec() {
...
     // Two-theta = polar angle = scattering angle = between +Z vector and the
     // scattered beam
```
Comment (SimonHeybrock) [inline Framework/Crystal/src/NormaliseVanadium.cpp]: use `twoTheta`

File: Framework/CurveFitting/src/Algorithms/ConvertToYSpace.cpp [hunk 11]
```diff
@@ -61,24 +61,23 @@ DetectorParams ConvertToYSpace::getDetectorParameters(
     throw std::invalid_argument(
         "ConvertToYSpace - Workspace has no source/sample.");
   }
-  Geometry::IDetector_const_sptr det;
-  try {
-    det = ws->getDetector(index);
-  } catch (Kernel::Exception::NotFoundError &) {
+
+  const auto &spectrumInfo = ws->spectrumInfo();
```
Comment (SimonHeybrock) [inline Framework/CurveFitting/src/Algorithms/ConvertToYSpace.cpp]: Consider passing `const SpectrumInfo &` to the function, instead of `ws`.

File: Framework/CurveFitting/src/Algorithms/VesuvioCalculateGammaBackground.cpp [hunk 12]
```diff
@@ -542,26 +540,30 @@ void VesuvioCalculateGammaBackground::cacheInstrumentGeometry() {
   m_foils1.resize(nfoils);
   for (size_t i = 0; i < nfoils; ++i) {
     const auto &foil0 = foils0[i];
+    if (!foil0)
+      throw std::invalid_argument("Error - foil component is NULL.");
```
Comment (SimonHeybrock) [inline Framework/CurveFitting/src/Algorithms/VesuvioCalculateGammaBackground.cpp]: Why this change?

File: Framework/CurveFitting/src/Functions/FunctionQDepends.cpp [hunk 13]
```diff
@@ -136,9 +136,13 @@ std::vector<double> FunctionQDepends::extractQValues(
     size_t numHist = workspace.getNumberHistograms();
     for (size_t wi = 0; wi < numHist; wi++) {
       try {
```
Comment (SimonHeybrock) [inline Framework/CurveFitting/src/Functions/FunctionQDepends.cpp]: Remove the `try` and `catch`, simply use

```cpp
if (!spectrumInfo.hasDetectors(wi)) {
  g_log.debug("Cannot populate Q values from workspace");
  qs.clear();
  break;
}

**Review by SimonHeybrock — COMMENTED** [review]
Some comment below:

File: Framework/Crystal/src/AnvredCorrection.cpp (hunk 3 above)
Comment (PranavBahuguna) [inline Framework/Crystal/src/AnvredCorrection.cpp]: Can you clarify? I'm unsure what you are referring to.

File: Framework/CurveFitting/src/Algorithms/VesuvioCalculateGammaBackground.cpp (hunk 12 above)
Comment (PranavBahuguna) [inline Framework/CurveFitting/src/Algorithms/VesuvioCalculateGammaBackground.cpp]: In `ConverrtToYSpace::getComponentParameter`, I was forced to pass an `IComponent` rather than `IComponent_const_sptr` to pass a detector to the method. There used to be null pointer check in this method, but as that could no longer be done there, I moved that check to be done outside the method on a detector before it is passed in.

It is an imperfect solution, so let me know if there are any better ways of going about this.

File: Framework/Crystal/src/AnvredCorrection.cpp (hunk 3 above)
Comment (SimonHeybrock) [inline Framework/Crystal/src/AnvredCorrection.cpp]: I though the variable was unused now, but actually it is still used, so ignore my comment.

File: Framework/CurveFitting/src/Algorithms/VesuvioCalculateGammaBackground.cpp (hunk 12 above)
Comment (SimonHeybrock) [inline Framework/CurveFitting/src/Algorithms/VesuvioCalculateGammaBackground.cpp]: Checking above:
```cpp
auto foils0 = inst->getAllComponentsWithName("foil-pos0");
auto foils1 = inst->getAllComponentsWithName("foil-pos1");
```
I think `Instrument` does not return `nullptr` in this vectors, so the check is not necessary?

No, there is no way of obtaining a shared pointer to a detector, since we intend to remove that mechanism altogether.

Comment (PranavBahuguna) [inline Framework/CurveFitting/src/Algorithms/VesuvioCalculateGammaBackground.cpp]: Alright, I'll remove it.

File: Framework/CurveFitting/src/Algorithms/ConvertToYSpace.cpp (hunk 11 above)
Comment (PranavBahuguna) [inline Framework/CurveFitting/src/Algorithms/ConvertToYSpace.cpp]: We retrieve several other parameters from ws though, I'm not sure this is really necessary.

Comment (SimonHeybrock) [inline Framework/CurveFitting/src/Algorithms/ConvertToYSpace.cpp]: Ok.

File: Framework/Crystal/src/AnvredCorrection.cpp [hunk 14]
```diff
@@ -266,10 +262,9 @@ void AnvredCorrection::execEvent() {
   Instrument_const_sptr inst = m_inputWS->getInstrument();
 
   const V3D samplePos = inst->getSample()->getPos();
```
Comment (SimonHeybrock) [inline Framework/Crystal/src/AnvredCorrection.cpp]: unused?

File: Framework/Crystal/src/AnvredCorrection.cpp [hunk 15]
```diff
@@ -261,8 +261,6 @@ void AnvredCorrection::execEvent() {
   // If sample not at origin, shift cached positions.
   Instrument_const_sptr inst = m_inputWS->getInstrument();
```
Comment (SimonHeybrock) [inline Framework/Crystal/src/AnvredCorrection.cpp]: Also unused?

Comment (PranavBahuguna) [inline Framework/Crystal/src/AnvredCorrection.cpp]: No those are used in the `scale_init` method

**Comment by SimonHeybrock** [comment]
:shipit: once Jenkins gives his ok.
